### DISCUSSION TITLE
Coulomb force

### DISCUSSION THREAD

**Discussion by BertJorissen** [ScicraftLearn/Minelabs#486] [General]
Add the coulomb-force to charged particles.
A first try is made to implement this, but it refreshes the electric field every tick. This can be made more efficient by only changing the field when there is a change in the world.

- [x] Make a new type of block, chargedBlock, that has the property 'charge' and the 3D-vector 'ElectricField'
- [ ] Upon addition of a charged block, the block looks for all the charged blocks in a certain radius [different from current implementation]
- [ ] Change the ElectricField of each block on addition/removal/displacement of a block [different from current implementation]
- [x] If the block has a large ElectricField, move the block in the direction of the Coulomb Force
- [ ] Additional: change the behavior of the Coulomb force to a more stochastic manner. For example, if the field allong x and y are of the same magnitude, have 50% chance to move in each direction.
- [x] Make an animation, that you really see the block shifting from one position to the next.
- [ ] Make an electric-field-sensor. This would be a block with 'outside charge' Q=0. It would only calculate the electric field from the surrounding blocks like it would have charge Q=1, but it won't change the electric fields of the other blocks. This block would appear like an arrow, pointing in the direction of the electric field. The length of the arrow is dependend on the size of the electric field.
- [ ] Try to give 'velocity' to the block, depending on the Coulomb Force. Several ways are possible, wait more ticks to move a block if it has a lower ElectricField, or let the block move further if it has a higher field.
- [ ] Come up with a completely different, more efficient implementation of the coulomb force. The hard part is that the force is modeled in a discrete square 3D grid, with a discrete time step (tick) that varies. Also, try to implement something that is manageable for the server. An idea would be to maintain a list of all the charged blocks for each chunck, such that there isn't need for a sweep of each block with a certain radious.

Also have a look on the Wiki under [Coulomb Force](https://wiki.scicraft.be/index.php/Coulomb_Force).
Current branch: Workshop.
For more information, contact @BertJorissen or via Discord _BertJorissen#4688_ .

**Comment by pixar02** [comment]
Could you show an example? (how it was done in ElementaryCraft)
Having just a block with a charge should be fine, but if you want others to have it to you want a superclass that has all this so you can just override/add more to a different block

**Comment by BertJorissen** [comment]
_My apologies for the late reply._
I see that you already have an abstract block implemented on your fork.
In essence, in ElementaryCraft, I tried to make it as abstract as possible.

- Every element has the 'charge' property as an NBT-tag. Also, there is a Minecraft-tag with all the blocks that are 'charged'.
- When a charged block is placed somewhere, it looks in the neighborhood to other blocks with the 'charge'-tag. If this block has this tag, it calculates the distance to that block. Then, the difference in electric field is calculated.
- The Coulomb force is linear, in the sense that two block give two times the force. This isn't always the case in physics, but we can use this nice property.  So, when a block is added, it starts with the electrical field of `E=0`. It then loops over all the blocks in the neighborhood, and adds the appropriate electric field for the charge of the other block at that distance, `E+=Q/r²`, with `Q` the charge of the other block. Also, this other block gets the charge added to it of the value `E'+=q/r²`, with `q` the charge of the new block.
- During this loop, the new block will get the additional NBT-tag of _immovable_, such that the procedure of obtaining the electrical field from the previous step is still running even though is was already displaced by an other block.
- When the electrical field is updated, the force will be calculated. If this force is large enough (in a certain direction), the block will be destroyed and placed in the other direction. Upon destruction of the block, the same procedure as the placement of the block is done, only with `E-=Q/r²` and  `E'-=q/r²`. Up to a small numerical factor, this would mean that the electric field will return to zero if there are no more block is the neighborhood. Also, because the block updates it's value when it is moved, a small inaccuracy in the electric field would be restored when the block is moved. This however will mean that when an electric charge is _static_, the electric field can potentially take on values that are too large or too small, if the Minecraft server just happens to delete the block without running the right procedures.

Because the last version of ElementaryCraft was made using these _CodeBlocks_, this was way too complicated. Afterwards I realized that this could also be done using just a java-code, but then I had it already implemented.

I'm just not sure if this is the best way of implementing the Coulomb force, there are a lot of unnecessary check and redundant calculations. 

On the other hand, it is good to calculate the electric field, and only calculate the force afterwards, because this saves some computational time.

I'll try to work on an example in the next few days.
